### PR TITLE
feat: add `payload` and `add-signature` commands.

### LINKED ISSUE #205: TUF CLI: Improve experience by allowing offline signature generation
This would help using the TUF CLI in more tightly controlled key environments, e.g. an air-gapped situation.

I propose two methods:

`tuf payload <role>` that outputs the canonicalized, signable content of the role's metadata file
`tuf add-signature <role> --signature <sig_file>` that adds a signature (hex-encoded bytes), generated by an offline tool, to the role. This should verify the signature first.

@haydentherapper

### PR DESCRIPTION
**Pull request by znewman01**
Fixes #205.

### CODE REVIEW

**Comment by znewman01** [comment]
ping @asraa (not to nag/rush, just want to make sure it's on your radar)

**Comment by znewman01** [comment]
Oh also CC @haydentherapper

File: repo.go
Comment (rdimitrov) [inline repo.go:1102]: It's a small nit, but I noticed the `tuf payload` command implicitly expects a role filename instead of what's in its description - i.e., `root.json` instead of `root`. Adding a test for that would be helpful too 👍

Comment (znewman01) [reply repo.go:1116]: Added a test that we get `ErrMissingMetadata` and changed help text to expect `<metadata>` instead of `<role>`

File: cmd/tuf/add_signature.go
Comment (znewman01) [reply cmd/tuf/add_signature.go]: Good call--this is sort-of meaningless without the "offline tool" implemented.

I've added `sign-payload` and verified that the `payload` -> `sign-payload` -> `add-signatures` flow works. I've also clarified the help text here to be agnostic to the format; the requirement is just that it comes from `sign-payload`.

**Comment by znewman01** [comment]
I've pushed a couple commits to fix the issues mentioned; of note, I add a `sign-payload` command which is the "offline tool" we've been dancing around.

File: repo.go
Comment (rdimitrov) [inline repo.go:526]: Maybe use something better than -1?

File: repo.go
Comment (rdimitrov) [inline repo.go:531]: Same here 👍

File: repo.go
Comment (rdimitrov) [inline repo.go:538]: Same here too 👍

**Review by rdimitrov — COMMENTED** [review]
I'd also suggest updating the list of commands in the README.md file by adding their descriptions along with examples of how to use them 👍

File: repo.go
Comment (znewman01) [reply repo.go:530]: Happy to change this, but my understanding is that in Go it's [idiomatic to totally ignore the success-case return value](https://stackoverflow.com/questions/46726832/how-to-return-int-or-nil-in-golang) if an error is passed. In fact, I see [Go by Example](https://gobyexample.com/errors) does exactly this, with a `-1` return value.

What do you propose I do instead?

I've improved the documentation for this function to be clearer about what it returns in these cases.

File: README.md
Comment (asraa) [inline README.md:268]: Could you point out where the keys should be stored if you are doing this? e.g. this expected a folder located like `keys/root.json` containing key information (output by gen-key)

File: repo.go
Comment (asraa) [inline repo.go:1090]: should this make sure this is json canonicalized?
it seems like it's just json unmarshalled into data.Signed
https://github.com/theupdateframework/go-tuf/blob/0e25bb50a25b479318f82432dc61ec5461c28322/repo.go#L675-L678

It seems like in most places we set the meta to be the canonicalized form (https://github.com/theupdateframework/go-tuf/blob/0e25bb50a25b479318f82432dc61ec5461c28322/repo.go#L511) but what if someone uses a local store and edits the metadata retrieved?

Maybe I should open a separate issue about this, because I'm not convinced we canonicalize consistently.

**Review by asraa — APPROVED** [review]
Just some minor comments!

File: repo.go
Comment (znewman01) [reply repo.go:530]: > and I wonder if having that number of keys returned is necessary? 

I think it might be useful for programmatic callers. I might want to double check that I've signed with the same number of keys that I expected.

> As far as I see we only use this value for logging a message, so we should be safe to return only the err value and log numKeys again whenever it is necessary from within the SignPayload() function. What are your thoughts on that?

I'd strongly prefer to avoid printing messages that we expect users to see within the library calls. Debug messages etc. are reasonable, but the output of the command should come from the command itself. Otherwise, applications might see our output in the middle of their operation.

So my vote is to leave it as-is. I would also be willing remove the number of keys from the output and the return value if you feel very strongly.

File: repo.go
Comment (znewman01) [reply repo.go:1090]: See my comments in #247. I do think you're broadly right.

That said, I think there's enough in-flight at the moment that we should probably open a separate issue and figure out a consistent strategy for canonicalization there. I'll go ahead and do that.

Comment (znewman01) [reply repo.go:1090]: https://github.com/theupdateframework/go-tuf/issues/261

File: repo.go
Comment (rdimitrov) [reply repo.go:530]: > I think it might be useful for programmatic callers. I might want to double check that I've signed with the same number of keys that I expected.

I'm okay with keeping it then if there's such a use case 👍 

> I'd strongly prefer to avoid printing messages that we expect users to see within the library calls. Debug messages etc. are reasonable, but the output of the command should come from the command itself. Otherwise, applications might see our output in the middle of their operation.
> 
Agree 👍 Makes sense.

**Review by rdimitrov — APPROVED** [review]

**Comment by znewman01** [comment]
No substantial changes but looks like I need a re-review, @asraa 

Otherwise I think this is good to go except I need one more approval from a maintainer—@trishankatdatadog any chance you could give a quick review?

**Comment by trishankatdatadog** [comment]
> Otherwise I think this is good to go except I need one more approval from a maintainer—@trishankatdatadog any chance you could give a quick review?

Sorry, no time 😕 Could someone else from DD take over (@ethan-lowman-dd @hosseinsia)?

File: README.md
Comment (asraa) [inline README.md:135]: FYI sorry about the badgering about the canonicalization again but if https://github.com/theupdateframework/go-tuf/pull/247/files?file-filters%5B%5D=.go&show-viewed-files=true is merged first (trynna get that in...), then this won't hold true -- metadata to the store won't be written in canonical form ahead of time.

Currently the tests use `SignPayload` which will internally canonicalize -- do you think you could add a test demonstrating that Payload() is canonical? Maybe show Canonicalize is no-op to the output

**Review by asraa — DISMISSED** [review]

File: README.md
Comment (znewman01) [reply README.md:135]: Ha, fair. Added [ee3e8e](https://github.com/theupdateframework/go-tuf/pull/214/commits/ee3e8ee3c70b0c35316338f53b1ab4afb66e396e) to check. (Turns out that `cjson.EncodeCanonical` is not idempotent!)

**Comment by znewman01** [comment]
Rebased and added a commit [ee3e8e](https://github.com/theupdateframework/go-tuf/pull/214/commits/ee3e8ee3c70b0c35316338f53b1ab4afb66e396e) that adds a test and fixes it in light of #247

**Review by rdimitrov — APPROVED** [review]
LGTM 👍

File: repo.go
Comment (cedricvanrompay-datadog) [inline repo.go:527]: Nit: I know this was already the case before this PR, but isn't it weird that the error is named "insufficient keys" when it is only returned when *no keys* where found? Shouldn't it be called `ErrNoKeys`? I would expect an "insufficient keys" errors to be returned when the number of keys we found is lesser than the threshold for this role.

**Review by cedricvanrompay-datadog — APPROVED** [review]
On behalf of Datadog: looks good to me.

I agree that [it would be better to make sure the data we sign is properly canonicalised](https://github.com/theupdateframework/go-tuf/pull/214#discussion_r840653371), I saw that a dedicated issue has been created.

**Comment by cedricvanrompay-datadog** [comment]
@asraa @trishankatdatadog approved on behalf of Datadog

File: repo_test.go
Comment (cedricvanrompay-datadog) [inline repo_test.go:1922]: Shouldn't we test that a client would accept the generated output?

File: repo.go
Comment (znewman01) [reply repo.go]: Renamed `ErrInsufficientKeys` to `ErrNoKeys`

File: README.md
Comment (znewman01) [reply README.md]: Updated the documentatation

**Comment by znewman01** [comment]
Need quick re-reviews again

**Review by rdimitrov — DISMISSED** [review]

**Comment by trishankatdatadog** [comment]
> Need quick re-reviews again

@cedricvanrompay-datadog would you pls mind reviewing?

On that note, I think it's worth nominating @znewman01 and @cedricvanrompay-datadog as additional maintainers. What do the @theupdateframework/go-tuf-maintainers think?

**Review by asraa — DISMISSED** [review]
Thanks for the test that `Payload` is canonicalized!

**Comment by asraa** [comment]
> On that note, I think it's worth nominating @znewman01 and @cedricvanrompay-datadog as additional maintainers. What do the https://github.com/orgs/theupdateframework/teams/go-tuf-maintainers think?

Def +1 to Zack, solid reviews, issue filing, and PRs.

**Comment by znewman01** [comment]
@cedricvanrompay-datadog it won't let me re-request a review for you for some reason so tagging you in manually

**Comment by znewman01** [comment]
Gentle ping @cedricvanrompay-datadog @asraa

**Review by asraa — APPROVED** [review]

**Pull request merged**
